### PR TITLE
Support update encoder settings during encoding

### PR DESCRIPTION
**Pull request by tongyuantongyu**
This is the first part for #761.

@y-guyon, I'm using `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS` due to the following considerations, but I'm open to change:

- We want to update settings in `avifEncoder.csOptions` as well, but it's an kv map emulated using array, which is hard to do comparison.
- This avoids changing libavif's current behavior.

### CODE REVIEW

**Comment by y-guyon** [comment]
> This is the first part for #761.

Thanks for doing that and also for adding a good unit test.

> @y-guyon, I'm using `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS` due to the following considerations, but I'm open to change

I would rather avoid adding a flag if there is no need to. I agree it adds a few functions and it requires a backup of `avifEncoder` fields at every `avifEncoderAddImage()` call, so up to you.

> * We want to update settings in `avifEncoder.csOptions` as well, but it's an kv map emulated using array, which is hard to do comparison.

```c
avifBool avifEncoderChanged(const avifEncoder* before, const avifEncoder* after) {
    if (before->speed != after->speed ||
        ... ||
        before->csOptions.count != after->csOptions.count) {
        return AVIF_TRUE;
    }
    for (uint32_t i = 0; i < csOptions->count; ++i) {
        avifCodecSpecificOption * entryBefore = &before->csOptions.entries[i];
        avifCodecSpecificOption * entryAfter = &after->csOptions.entries[i];
        if (strcmp(entryBefore->key, entryAfter->key) || strcmp(entryBefore->value, entryAfter->value)) {
            // This could be a false positive if options got reordered but it is unlikely and it is not a big issue.
            return AVIF_TRUE;
        }
    }
    return AVIF_FALSE;
}
```

`avifCodecSpecificOptionsSet()` has a similar behavior.

> * This avoids changing libavif's current behavior.

It's probably fine. I would be surprised if a single user changed `avifEncoder` fields after `avifEncoderAddImage()` and expected them not to be taken into account.

File: tests/gtest/avifchangesettingtest.cc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  ASSERT_EQ(avifEncoderFinish(encoder.get(), &encodedAvif), AVIF_RESULT_OK);
+
+  // Decode
+  testutil::AvifDecoderPtr decoder(avifDecoderCreate(), avifDecoderDestroy);
+  ASSERT_NE(decoder, nullptr);
+
+  // The second frame is set to have far better quality,
+  // and should be much bigger, so small amount of data at beginning
+  // should be enough to decode the first frame.
+  auto io = testutil::AvifIOCreateLimitedReader(
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/avifchangesettingtest.cc]: Done

File: tests/gtest/aviftest_helpers.h [hunk 2]
```diff
@@ -61,6 +61,16 @@ bool AreImagesEqual(const avifImage& image1, const avifImage& image2,
 
 //------------------------------------------------------------------------------
 
+struct AvifIOLimitedReader {
+  static constexpr uint64_t NoClamp = UINT64_MAX;
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/aviftest_helpers.h]: Done

[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/aviftest_helpers.h]: Done

File: tests/gtest/aviftest_helpers.cc [hunk 3]
```diff
@@ -211,6 +211,38 @@ bool AreImagesEqual(const avifImage& image1, const avifImage& image2,
   return true;
 }
 
+static avifResult avifIOLimitedReaderRead(struct avifIO* io, uint32_t readFlags,
+                                          uint64_t offset, size_t size,
+                                          avifROData* out) {
+  auto reader = (AvifIOLimitedReader*)io;
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/aviftest_helpers.cc]: Done

File: tests/gtest/avifchangesettingtest.cc [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "avif/avif.h"
+#include "aviftest_helpers.h"
+#include "gtest/gtest.h"
+
+namespace libavif {
+namespace {
+
+TEST(ChangeSettingTest, AOM) {
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/avifchangesettingtest.cc]: `TEST_P` seems to test against different combinations of parameters. A helper function with some parameters looks more suitable here.

File: src/codec_aom.c [hunk 5]
```diff
@@ -689,41 +694,51 @@ static avifResult aomCodecEncodeImage(avifCodec * codec,
...
 
-        if (!avifProcessAOMOptionsPreInit(codec, alpha, &cfg)) {
+        if (!avifProcessAOMOptionsPreInit(codec, alpha, cfg)) {
             return AVIF_RESULT_INVALID_CODEC_SPECIFIC_OPTION;
         }
 
         aom_codec_flags_t encoderFlags = 0;
         if (image->depth > 8) {
             encoderFlags |= AOM_CODEC_USE_HIGHBITDEPTH;
         }
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply src/codec_aom.c]: Done.

`avifEncoderAddImage` calls `avifEncoderAddImageInternal`, which already ensured subsequent images has same metadata with the first one at the very beginning "Validate images" section.

**Comment by tongyuantongyu** [comment]
I now agree with you.

For codec specific options, I realized user can only change it using `avifEncoderSetCodecSpecificOption()`. We can then simply record if this function was called by user, and the complicated comparison can be avoid then.

I was thinking about some users accidentally changed encoder settings, and found the program start to fail (when they are using rav1e or svt). But you are right this is certainly a bug and user should correct their usage.

From comment of `ResizePendingParams` in libaom, `AOME_SET_SCALEMODE` is reset after encoding each frame. I'm not sure if there are other options work in this way, so I kept `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS` as a way to explicitly request reconfiguring.
```c
/*!
 * \brief Desired dimensions for an externally triggered resize.
 *
 * When resize is triggered externally, the desired dimensions are stored in
 * this struct until used in the next frame to be coded. These values are
 * effective only for one frame and are reset after they are used.
 */
typedef struct {
  int width;  /*!< Desired resized width */
  int height; /*!< Desired resized height */
} ResizePendingParams;
```

File: src/codec_aom.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (y-guyon) [reply src/codec_aom.c]: > `avifEncoderAddImage` calls `avifEncoderAddImageInternal`, which already ensured subsequent images has same metadata with the first one at the very beginning "Validate images" section.

Thank you for taking a look.
Unfortunately the code you are referring to only makes sure all cells in a layer have the same features as the first cell, but they are not compared with the first layer.

https://github.com/AOMediaCodec/libavif/blob/1ce971ee6de4120a7607da8fec882879de9ae546/src/write.c#L642-L649

**Review by y-guyon — COMMENTED** [review]
> From comment of `ResizePendingParams` in libaom, `AOME_SET_SCALEMODE` is reset after encoding each frame. I'm not sure if there are other options work in this way, so I kept `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS` as a way to explicitly request reconfiguring.

Good catch. I still think libavif should have a consistent behavior, so why not remove `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS` for simplicity and apply all codec flags at every `avifEncoderAddImage()` (or at least the ones we know will be reset by the codec)? The user can then remove them before doing another `avifEncoderAddImage()` if needed. The fact that it is not documented anywhere in avif.h makes me think it is more of a bug or oversight than a feature.

On the other hand, before this PR there was no way to modify a codec flag and take it into account past the first `avifEncoderAddImage()` call. I also wonder how bad calling `avifProcessAOMOptionsPreInit()` after `aom_codec_enc_init()` can be.

On top of that, the current documentation says that `avifEncoderSetCodecSpecificOption()` should not be called after `avifEncoderAddImage()`:

https://github.com/AOMediaCodec/libavif/blob/bbf459eec9a75a20010f5251a9575dfe1717885b/include/avif/avif.h#L1117-L1123

This does not seem like a hard rule, just discouraged.

I can think of these options:

- Make it a hard rule, and return `AVIF_RESULT_CANNOT_CHANGE_SETTING` if `avifEncoderSetCodecSpecificOption()` is called after `avifEncoderAddImage()`. This may be a limiting behavior but not a breaking change. No need for `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS`.
- Remove the rule, update all/some codec flags at every `avifEncoderAddImage()`, and make sure everything works as intended. No need for `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS`. This may require more work, especially testing, likely in another PR.
- Remove the rule, discard all codec flags at every `avifEncoderAddImage()`, and apply new ones added with `avifEncoderSetCodecSpecificOption()` at next `avifEncoderAddImage()`. No need for `AVIF_ADD_IMAGE_FLAG_UPDATE_SETTINGS`. This is closer to the libaom API spirit in my opinion, but slightly more obscure on the libavif side. It should be easier to implement as the burden of understanding libaom flags is forwarded to the libavif user, but it will require a lengthy and clear explanation as a comment in `avif.h`.

What do you think? Personally I would lean toward the last option.

**Comment by tongyuantongyu** [comment]
> `avifProcessAOMOptionsPreInit()`

This function only set up value of `cfg.rc_end_usage`.  Naming it "PreInit" is because `aom_codec_enc_init()` uses `aom_codec_enc_cfg` so we need to set up it before the call. `aom_codec_enc_config_set()` is meant to update options in `aom_codec_enc_cfg` so this is totally fine and intended. We may rename `avifProcessAOMOptionsPreInit()` to `avifProcessAOMOptionsSetupCfg()` and `avifProcessAOMOptionsPostInit()` to `avifProcessAOMOptionsSetupExtraCfg()` (or something like these) to better describe what they do.

> current documentation says

Thanks for remind this. I believe we should update the document, as we would definitely want to allow some options, `cq-level` for example, to be changed during encoding.

> What do you think?

After reading AOM code again, I found that `AOME_SET_SCALEMODE` cannot to be set by `aom_codec_set_option()` (See `av1/av1_cx_iface.c#encoder_set_option()`), and those options that can be set by `aom_codec_set_option()` can also be set using command line flags when encoding AV1 video, which clearly won't make sense if they only affects the first frame. So I'd assume that they are persistent and `AOME_SET_SCALEMODE` is the special case. We probably won't put scale mode into `csOptions`, so I believe the second choice should be fine, and there should be no need to list any special cases for now.

File: include/avif/avif.h [hunk 6]
```diff
@@ -146,7 +146,8 @@ typedef enum avifResult
     AVIF_RESULT_WAITING_ON_IO, // similar to EAGAIN/EWOULDBLOCK, this means the avifIO doesn't have necessary data available yet
     AVIF_RESULT_INVALID_ARGUMENT, // an argument passed into this function is invalid
     AVIF_RESULT_NOT_IMPLEMENTED,  // a requested code path is not (yet) implemented
-    AVIF_RESULT_OUT_OF_MEMORY
+    AVIF_RESULT_OUT_OF_MEMORY,
+    AVIF_RESULT_CANNOT_CHANGE_SETTING, // a setting that can't change is changed during encoding
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply include/avif/avif.h]: done

File: src/avif.c [hunk 7]
```diff
@@ -96,6 +96,7 @@ const char * avifResultToString(avifResult result)
         case AVIF_RESULT_INVALID_ARGUMENT:              return "Invalid argument";
         case AVIF_RESULT_NOT_IMPLEMENTED:               return "Not implemented";
         case AVIF_RESULT_OUT_OF_MEMORY:                 return "Out of memory";
+        case AVIF_RESULT_CANNOT_CHANGE_SETTING:             return "Can not change some settings during encoding";
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply src/avif.c]: done

File: tests/gtest/avifchangesettingtest.cc [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+TEST(ChangeSettingTest, SVT) {
+  TestEncodeDecode(AVIF_CODEC_CHOICE_SVT, {}, false, false);
+}
+
+TEST(ChangeSettingTest, ChangeCsOptions) {
+  TestEncodeDecode(AVIF_CODEC_CHOICE_AOM, {}, true, true);
+}
+
+TEST(ChangeSettingTest, UnchangableSetting) {
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/avifchangesettingtest.cc]: done

File: tests/gtest/avifchangesettingtest.cc [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+TEST(ChangeSettingTest, RAV1E) {
+  TestEncodeDecode(AVIF_CODEC_CHOICE_RAV1E, {}, false, false);
+}
+
+TEST(ChangeSettingTest, SVT) {
+  TestEncodeDecode(AVIF_CODEC_CHOICE_SVT, {}, false, false);
+}
+
+TEST(ChangeSettingTest, ChangeCsOptions) {
+  TestEncodeDecode(AVIF_CODEC_CHOICE_AOM, {}, true, true);
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply tests/gtest/avifchangesettingtest.cc]: done

File: src/write.c [hunk 10]
```diff
@@ -317,6 +319,47 @@ void avifEncoderDestroy(avifEncoder * encoder)
 void avifEncoderSetCodecSpecificOption(avifEncoder * encoder, const char * key, const char * value)
 {
     avifCodecSpecificOptionsSet(encoder->csOptions, key, value);
+    encoder->data->csOptionsUpdated = AVIF_TRUE;
+}
+
+avifBool avifEncoderCheckSettingsChange(avifEncoder * encoder, avifAddImageFlag * flags)
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply src/write.c]: done

[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply src/write.c]: done

File: src/codec_aom.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply src/codec_aom.c]: Thanks for your careful check. The validation is indeed missing, and I'm adding it.

File: src/write.c [hunk 11]
```diff
@@ -649,8 +656,11 @@ static avifResult avifEncoderAddImageInternal(avifEncoder * encoder,
         return AVIF_RESULT_NO_CODEC_AVAILABLE;
     }
 
-    if (!avifEncoderCheckSettingsChange(encoder, &addImageFlags)) {
+    avifBool updateConfig = AVIF_FALSE;
+    if (!avifEncoderSettingsChanged(encoder, &updateConfig)) {
         return AVIF_RESULT_CANNOT_CHANGE_SETTING;
+    } else {
```
Comment (y-guyon) [inline src/write.c:662]: No need for `else`

File: src/write.c [hunk 12]
```diff
@@ -853,10 +863,23 @@ static avifResult avifEncoderAddImageInternal(avifEncoder * encoder,
...
+        // subsequence images must have alpha as well.
+        if ((imageMetadata->width != firstCell->width) || (imageMetadata->height != firstCell->height) ||
+            (imageMetadata->depth != firstCell->depth) || (imageMetadata->yuvFormat != firstCell->yuvFormat) ||
+            (imageMetadata->yuvRange != firstCell->yuvRange) || (imageMetadata->colorPrimaries != firstCell->colorPrimaries) ||
+            (imageMetadata->transferCharacteristics != firstCell->transferCharacteristics) ||
+            (imageMetadata->matrixCoefficients != firstCell->matrixCoefficients) ||
+            (!!imageMetadata->alphaPlane != !!firstCell->alphaPlane) ||
+            (imageMetadata->alphaPremultiplied != firstCell->alphaPremultiplied) ||
+            (encoder->data->alphaPresent && !firstCell->alphaPlane)) {
+            return AVIF_RESULT_INCOMPATIBLE_IMAGE;
```
Comment (y-guyon) [inline src/write.c:882]: Thanks for handling that!

**Review by y-guyon — APPROVED** [review]

File: include/avif/avif.h [hunk 13]
```diff
@@ -148,6 +148,7 @@ typedef enum avifResult
     AVIF_RESULT_NOT_IMPLEMENTED,  // a requested code path is not (yet) implemented
     AVIF_RESULT_OUT_OF_MEMORY,
     AVIF_RESULT_CANNOT_CHANGE_SETTING, // a setting that can't change is changed during encoding
+    AVIF_RESULT_INCOMPATIBLE_IMAGE     // given image is not compatible with already encoded image
```
[the remark this answers is not in the document]
Comment (tongyuantongyu) [reply include/avif/avif.h:151]: Understand. If both are newly added, I would make `AVIF_RESULT_INVALID_IMAGE_GRID` and `AVIF_RESULT_INCOMPATIBLE_IMAGE` share one value, but `AVIF_RESULT_INVALID_IMAGE_GRID` is already there and is very specific about grid.

**Review by wantehchang — COMMENTED** [review]
Hi Yuan:

Thank you for the pull request and I am sorry about the late review. I reviewed this pull request while trying to import a new libavif snapshot into Google's internal repository today.

Regarding the rule for when `avifEncoderSetCodecSpecificOption()` can be called, the third option (quoted below) seems the best to me:

> - Remove the rule, discard all codec flags at every `avifEncoderAddImage()`, and apply new ones added with `avifEncoderSetCodecSpecificOption()` at next `avifEncoderAddImage()`.

It seems that you did not implement this option. Instead, you implemented the second option (quoted below):

> - Remove the rule, update all/some codec flags at every `avifEncoderAddImage()`, and make sure everything works as intended.

Did I understand it correctly?

**Pull request merged**